lib/libzfs/libzfs_sendrecv.c

```diff
@@ -3755,7 +3887,8 @@ zfs_receive_impl(libzfs_handle_t *hdl, const char *tosnap,
 		assert(DMU_GET_STREAM_HDRTYPE(drrb->drr_versioninfo) ==
 		    DMU_COMPOUNDSTREAM);
 		return (zfs_receive_package(hdl, infd, tosnap, flags, &drr,
```

Comment:
A lot of the libzfs_sendrecv code is hard to trace. Most of it is because of the way that -R and -I were implemented, and the "compound stream" idea.  If you modify zfs_receive_one() to take a props nvlist, and combine that with the property nvlist handling code already present, that should work.